core/src/test/java/com/linecorp/armeria/client/proxy/HAProxyClientIntegrationTest.java

```diff
@@ -0,0 +1,393 @@
+                msgRef.set((HAProxyMessage) msg);
+                return;
+            }
+            final FullHttpRequest request = (FullHttpRequest) msg;
+            final DefaultFullHttpResponse response;
+            if ("PRI".equals(request.method().name())) {
+                // reject http2 preface
+                final HttpHeaders headers = new DefaultHttpHeaders().add(HttpHeaderNames.CONNECTION, "close");
+                response = new DefaultFullHttpResponse(HttpVersion.HTTP_1_1,
+                                                       HttpResponseStatus.NOT_IMPLEMENTED,
+                                                       Unpooled.EMPTY_BUFFER,
+                                                       headers,
+                                                       EmptyHttpHeaders.INSTANCE);
+            } else {
+                response = new DefaultFullHttpResponse(HttpVersion.HTTP_1_1,
+                                                       HttpResponseStatus.OK,
+                                                       Unpooled.copiedBuffer(strRepr(msgRef.get()),
+                                                                             StandardCharsets.US_ASCII));
+            }
+
+            ReferenceCountUtil.release(msg);
+            ctx.writeAndFlush(response);
+            ctx.close();
+        });
+
+        try (ClientFactory clientFactory =
+                     ClientFactory.builder()
+                                  .proxyConfig(ProxyConfig.haproxy(proxyAddr, srcAddr))
+                                  .useHttp2Preface(true)
+                                  .build()) {
+
+            final WebClient webClient = WebClient.builder(SessionProtocol.HTTP, destEndpoint)
+                                                 .factory(clientFactory)
+                                                 .decorator(LoggingClient.newDecorator())
+                                                 .build();
+            final CompletableFuture<AggregatedHttpResponse> responseFuture =
+                    webClient.get(PROXY_PATH).aggregate();
+            final AggregatedHttpResponse response = responseFuture.get(10, TimeUnit.SECONDS);
+            assertThat(response.status()).isEqualTo(HttpStatus.OK);
+            final String expectedResponse = String.format("%s-%s", srcAddr, destAddr);
+            assertThat(response.contentUtf8()).isEqualTo(expectedResponse);
+        }
+    }
+
+    private static class ProtocolEndpointProvider implements ArgumentsProvider {
+        @Override
+        public Stream<? extends Arguments> provideArguments(ExtensionContext context) {
+            return Stream.concat(
+                    SessionProtocol.httpValues().stream()
+                                   .map(p -> arguments(p, backendServer.httpSocketAddress())),
+                    SessionProtocol.httpsValues().stream()
+                                   .map(p -> arguments(p, backendServer.httpsSocketAddress()))
+            );
```

Comment:
Could simplify:
```java
Stream.concat(SessionProtocol.httpValues().stream(), SessionProtocol.httpsValues().stream())
      .map(p -> arguments(p, backendServer.httpSocketAddress()));
```

Reply:
Actually we also need to divide `backendServer.httpSocketAddress` vs `backendServer.httpsSocketAddress` 😅

Reply:
Ah... Sorry. I missed that 🤣